pkg/controller/pxc/replication.go

```diff
@@ -86,6 +97,259 @@ func (r *ReconcilePerconaXtraDBCluster) removeOutdatedServices(cr *api.PerconaXt
 	return nil
 }
 
+func (r *ReconcilePerconaXtraDBCluster) reconcileReplication(cr *api.PerconaXtraDBCluster) error {
+	if cr.Status.PXC.Ready < 1 || cr.Spec.Pause {
+		return nil
+	}
+
+	sfs := statefulset.NewNode(cr)
+
+	list := corev1.PodList{}
+	if err := r.client.List(context.TODO(),
+		&list,
+		&client.ListOptions{
+			Namespace:     cr.Namespace,
+			LabelSelector: labels.SelectorFromSet(sfs.Labels()),
+		},
+	); err != nil {
+		return errors.Wrap(err, "get pod list")
+	}
+
+	primary, err := r.getPrimaryPod(cr)
+	if err != nil {
+		return errors.Wrap(err, "get primary pxc pod")
+	}
+
+	for _, pod := range list.Items {
+		if pod.Status.PodIP == primary || pod.Name == primary {
+			primary = fmt.Sprintf("%s.%s.%s", pod.Name, sfs.Name(), cr.Namespace)
+			break
+		}
+	}
+
+	sort.Slice(list.Items, func(i, j int) bool {
+		return list.Items[i].Name > list.Items[j].Name
+	})
+
+	var primaryPod corev1.Pod
+	for _, pod := range list.Items {
+		if strings.HasPrefix(primary, fmt.Sprintf("%s.%s.%s", pod.Name, sfs.Name(), cr.Namespace)) {
+			primaryPod = pod
+			break
+		}
+	}
+
+	user := "operator"
+	port := int32(33062)
+
+	primaryDB, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, primaryPod.Name+"."+cr.Name+"-pxc."+cr.Namespace, port)
+	if err != nil {
+		return errors.Wrapf(err, "failed to connect to pod %s", primaryPod.Name)
+	}
+
+	defer primaryDB.Close()
+
+	dbVer, err := primaryDB.Version()
+	if err != nil {
+		return errors.Wrap(err, "failed to get current db version")
+	}
+
+	if minReplicationVersion.Compare(version.Must(version.NewVersion(dbVer))) < 0 {
+		r.logger(cr.Name, cr.Namespace).Info("Failed to reconcile replication. Unsupported mysql version", "minimum version", minReplicationVersion.String())
+		return nil
+	}
+
+	err = removeOutdatedChannels(primaryDB, cr.Spec.PXC.ReplicationChannels)
+	if err != nil {
+		return errors.Wrap(err, "remove outdated replication channels")
+	}
+
+	if len(cr.Spec.PXC.ReplicationChannels) == 0 {
+		return nil
+	}
+
+	// if primary pod is not a replica, we need to make it as replica, and stop replication on other pods
+	_, ok := primaryPod.Labels[replicationPodLabel]
+	if !ok {
+		primaryPod.Labels[replicationPodLabel] = "true"
+		err = r.client.Update(context.TODO(), &primaryPod)
+		if err != nil {
+			return errors.Wrap(err, "add label to main replica pod")
+		}
+		for _, pod := range list.Items {
+			if pod.Name == primaryPod.Name {
+				continue
+			}
+			if _, ok := pod.Labels[replicationPodLabel]; ok {
+				r.logger(cr.Name, cr.Namespace).Info("Replication pod has changed", "new replication pod", pod.Name)
+				db, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, pod.Name+"."+cr.Name+"-pxc."+cr.Namespace, port)
```

Comment:
should we close db on error?
```suggestion
				db, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, pod.Name+"."+cr.Name+"-pxc."+cr.Namespace, port)
				defer db.Close()
```

Reply:
no, we need to use defer only if error from connecting to db is nil